src/Microsoft.AspNetCore.SignalR.Redis/Internal/RedisSubscriptionManager.cs

```diff
@@ -16,57 +16,74 @@ private class SubscriptionData
             public readonly HubConnectionStore Connections = new HubConnectionStore();
         }
 
-        // TODO: Investigate "memory leak" entries never get removed
         private readonly ConcurrentDictionary<string, SubscriptionData> _subscriptions = new ConcurrentDictionary<string, SubscriptionData>(StringComparer.Ordinal);
+        private readonly SemaphoreSlim _lock = new SemaphoreSlim(1, 1);
 
         public async Task AddSubscriptionAsync(string id, HubConnectionContext connection, Func<string, HubConnectionStore, Task> subscribeMethod)
         {
-            var subscription = _subscriptions.GetOrAdd(id, _ => new SubscriptionData());
-
-            await subscription.Lock.WaitAsync();
+            await _lock.WaitAsync();
```

Comment:
A global lock wrapping all logic in add and remove subscription could easily become a bottleneck.

Reply:
Yep, that's the main concern with doing this fix this way. However, the lock should, I believe, only be needed during connection/disconnection, which means it's not on the hottest of hot paths.